src/helpers.cpp

```diff
@@ -1,37 +1,116 @@
 #include "vowpalwabbit/vw.h"
 #include <Rcpp.h>
 
-std::string check_data(Rcpp::List & vwmodel, std::string & data_path, std::string mode="train") {
-    Rcpp::List vwmodel_data = vwmodel["data"];
-    std::string valid_data_str;
-    // // DEBUG
-    // // Rcpp::Rcout << vwmodel["params"] << std::endl;
-    // Rcpp::Rcout << Rcpp::as<std::string>(vwmodel["dir"]) << std::endl;
-    // Rcpp::Rcout << Rcpp::as<std::string>(vwmodel_data["train"]) << std::endl;
-    // Rcpp::Rcout << data_path << std::endl;
-    // // END DEBUG
+#include <fstream>
+#include <stdio.h>
+#include <string.h>
+
+#include <RApiSerializeAPI.h>
+
+#include <R.h>
+#include <R_ext/Rdynload.h>
+
+extern "C" {
+#include "md5.h"
+}
+
+
+// Based on code from R digest package http://dirk.eddelbuettel.com/code/digest.html
+// Copyright (C) 2003 - 2016  Dirk Eddelbuettel <edd@debian.org>
+std::string md5sum(char * char_x, uint32_t nChar) {
+    char output[33+1];
+    int output_length = -1;
+    md5_context ctx;
+    output_length = 16;
+    unsigned char md5sum[16];
+    int j;
+    md5_starts( &ctx );
+    md5_update( &ctx, (uint8 *) char_x, nChar);
+    md5_finish( &ctx, md5sum );
+    memcpy(output, md5sum, 16);
     
-    std::string vwmodel_data_file;
-    if (mode == "test") {
-        vwmodel_data_file = Rcpp::as<std::string>(vwmodel_data["test"]);
-    } else {
-        vwmodel_data_file = Rcpp::as<std::string>(vwmodel_data["train"]);
-    }
+    for (j = 0; j < 16; j++)
+        sprintf(output + j * 2, "%02x", md5sum[j]);
     
-    if(data_path.empty()) {
-        if(vwmodel_data_file.empty()) {
-            Rcpp::stop("No data provided");
-        } else {
-            valid_data_str = Rcpp::as<std::string>(vwmodel["dir"]) + vwmodel_data_file;
+    std::string output_str(output);
+    
+    return(output_str);
+}
+
+Rcpp::CharacterVector check_data(Rcpp::List & vwmodel, std::string & valid_data_str, SEXP & data=R_NilValue, std::string mode="train",
+                       SEXP & namespaces=R_NilValue, SEXP & keep_space=R_NilValue,
```

Comment:
Why the `&` next to `SEXP`? By default the `SEXP` is a pointer, causing a reference update to occur. Why aren't these typed? 

See my remarks here on how to use Null in _C++_ parameters:

 https://stackoverflow.com/questions/43479211/allow-c-constants-to-be-a-default-function-parameter-using-rcpp-attributes/43480020#43480020

Reply:
Thank you! 
I misunderstood this logic and will rework it now.